src/renderer/generic.zig

```diff
@@ -0,0 +1,2873 @@
+                state.terminal.flags.dirty = .{};
+                state.terminal.screen.dirty = .{};
+                {
+                    var it = state.terminal.screen.pages.pageIterator(
+                        .right_down,
+                        .{ .screen = .{} },
+                        null,
+                    );
+                    while (it.next()) |chunk| {
+                        var dirty_set = chunk.node.data.dirtyBitSet();
+                        dirty_set.unsetAll();
+                    }
+                }
+
+                // Update our viewport pin
+                self.cells_viewport = viewport_pin;
+
+                break :critical .{
+                    .bg = self.background_color orelse self.default_background_color,
+                    .screen = screen_copy,
+                    .screen_type = state.terminal.active_screen,
+                    .mouse = state.mouse,
+                    .preedit = preedit,
+                    .cursor_style = cursor_style,
+                    .color_palette = state.terminal.color_palette.colors,
+                    .full_rebuild = full_rebuild,
+                };
+            };
+            defer {
+                critical.screen.deinit();
+                if (critical.preedit) |p| p.deinit(self.alloc);
+            }
+
+            // Build our GPU cells
+            try self.rebuildCells(
+                critical.full_rebuild,
+                &critical.screen,
+                critical.screen_type,
+                critical.mouse,
+                critical.preedit,
+                critical.cursor_style,
+                &critical.color_palette,
+            );
+
+            // Notify our shaper we're done for the frame. For some shapers,
+            // such as CoreText, this triggers off-thread cleanup logic.
+            self.font_shaper.endFrame();
+
+            // Acquire the draw mutex because we're modifying state here.
+            {
+                self.draw_mutex.lock();
+                defer self.draw_mutex.unlock();
+
+                // Update our background color
+                self.uniforms.bg_color = .{
+                    critical.bg.r,
+                    critical.bg.g,
+                    critical.bg.b,
+                    @intFromFloat(@round(self.config.background_opacity * 255.0)),
+                };
+            }
+        }
+
+        /// Draw the frame to the screen.
+        ///
+        /// If `sync` is true, this will synchronously block until
+        /// the frame is finished drawing and has been presented.
+        pub fn drawFrame(
+            self: *Self,
+            sync: bool,
+        ) !void {
+            // We hold a the draw mutex to prevent changes to any
+            // data we access while we're in the middle of drawing.
+            self.draw_mutex.lock();
+            defer self.draw_mutex.unlock();
+
+            // There's probably a more elegant way to do this...
```

Comment:
We may want to introduce a `drawFrameStart()/drawFrameEnd()` `GraphicsAPI` function so that we can defer stuff like this to the implementation rather than in generic.